File: apps/worker/app/models/blocks.py
"""
Blocks model - Cleaned and optimized schema
"""
from datetime import datetime
from typing import Optional, Dict, Any

from sqlalchemy import String, Text, DateTime, Integer, func, ForeignKey, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base


class Block(Base):
    """Blocks table - scraped content data (cleaned schema)"""
    __tablename__ = "blocks"
    
    # Primary key - using integer to match Payload
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    
    # External Reference
    external_id: Mapped[str] = mapped_column(
        String(255), 
        nullable=False, 
        unique=True,
        index=True,
        doc="Unique identifier from Savee.it"
    )
    
    # Relationships (get source_type/username via relationships)
    source_id: Mapped[int] = mapped_column(
        ForeignKey('sources.id'), 
        nullable=False, 
        index=True,
        doc="Source that discovered this block"
    )
    run_id: Mapped[int] = mapped_column(
        ForeignKey('runs.id'), 
        nullable=False, 
        index=True,
        doc="Run that scraped this block"
    )
    
    # Content Info
    url: Mapped[str] = mapped_column(
        Text, 
        nullable=False,
        doc="Content URL on Savee.it"
    )
    title: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Title of the content"
    )
    description: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Description of the content"
    )
    
    # Media information
    media_type: Mapped[str] = mapped_column(
        String(20), 
        nullable=True,
        doc="Media type: image, video, gif, unknown"
    )
    image_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Original image URL"
    )
    video_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Original video URL"
    )
    thumbnail_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Thumbnail URL"
    )
    original_source_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Original source URL before Savee"
    )
    
    # Status and Processing
    status: Mapped[str] = mapped_column(
        String(50), 
        default="pending",
        nullable=False,
        doc="Processing status: pending, fetched, scraped, uploaded, error"
    )
    
    # Metadata
    tags: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        doc="Content tags"
    )
    content_metadata: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        doc="Additional metadata"
    )
    
    # Storage
    r2_key: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="R2 storage key for uploaded media"
    )
    
    # Error handling
    error_message: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Error message if processing failed"
    )
    
    # Timestamps (standard Payload)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        doc="When this block was created"
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
        doc="When this block was last updated"
    )

    # Relationships
    source = relationship("Source")
    run = relationship("Run")

    def __repr__(self) -> str:
        return f"<Block(id={self.id}, external_id='{self.external_id}', media_type='{self.media_type}', status='{self.status}')>"

File: apps/cms/src/collections/Blocks.ts
import type { CollectionConfig } from "payload";

export const Blocks: CollectionConfig = {
  slug: "blocks",
  admin: {
    useAsTitle: "title",
    defaultColumns: [
      "preview",
      "title",
      "media_type", // Match database column name
      "origin", // Computed: home | pop | username
      "saved_by",
      "status",
      "source",
      "run",
      "createdAt",
    ],
    description: "Individual scraped content blocks from Savee.it",
  },
  access: {
    read: () => true,
    create: () => true,
    update: () => true,
    delete: () => true,
  },
  fields: [
    // Preview (admin-only UI field)
    {
      name: "preview",
      type: "ui",
      label: "Preview",
      admin: {
        components: {
          Cell: "@/components/BlockPreviewCell",
        },
      },
    },
    // Saved by usernames
    {
      name: "saved_by",
      type: "ui",
      label: "Saved By",
      admin: {
        components: {
          Cell: "@/components/BlockUsersCell",
        },
      },
    },
    // Origin (computed from source.sourceType / username)
    {
      name: "origin",
      type: "ui",
      label: "Savee User / Origin",
      admin: {
        description: "home | pop | username (computed)",
        components: {
          Cell: "@/components/BlockOriginCell",
        },
      },
    },
    // External Reference
    {
      name: "external_id", // Match database column name
      type: "text",
      required: true,
      unique: true,
      label: "External ID",
      admin: {
        description: "Unique identifier from Savee.it",
      },
    },

    // Relationships (source info available via relationships)
    {
      name: "source",
      type: "relationship",
      relationTo: "sources",
      required: true,
      label: "Source",
    },
    {
      name: "run",
      type: "relationship",
      relationTo: "runs",
      required: true,
      label: "Run",
    },
    {
      name: "savee_user",
      type: "relationship",
      relationTo: "savee_users",
      label: "Savee User",
      admin: {
        description: "SaveeUser profile for user content organization",
        position: "sidebar",
      },
    },
    

    // Content Info
    {
      name: "url",
      type: "text",
      required: true,
      label: "Content URL",
    },
    {
      name: "title",
      type: "text",
      label: "Title",
    },
    {
      name: "description",
      type: "textarea",
      label: "Description",
    },

    // Media Information
    {
      name: "media_type", // Match database column name
      type: "select",
      label: "Media Type",
      options: [
        { label: "Image", value: "image" },
        { label: "Video", value: "video" },
        { label: "GIF", value: "gif" },
        { label: "Unknown", value: "unknown" },
      ],
    },
    {
      name: "image_url", // Match database column name
      type: "text",
      label: "Image URL",
    },
    {
      name: "video_url", // Match database column name
      type: "text",
      label: "Video URL",
    },
    {
      name: "thumbnail_url", // Match database column name
      type: "text",
      label: "Thumbnail URL",
    },

    // Status and Processing
    {
      name: "status",
      type: "select",
      required: true,
      defaultValue: "pending",
      label: "Status",
      options: [
        { label: "Pending", value: "pending" },
        { label: "Fetched", value: "fetched" },
        { label: "Scraped", value: "scraped" },
        { label: "Uploaded", value: "uploaded" },
        { label: "Error", value: "error" },
      ],
    },

    // Comprehensive OpenGraph Metadata
    {
      name: "og_title", // Match database column name
      type: "text",
      label: "OG Title",
      admin: {
        description: "OpenGraph title from meta tags",
      },
    },
    {
      name: "og_description", // Match database column name
      type: "textarea",
      label: "OG Description",
      admin: {
        description: "OpenGraph description from meta tags",
      },
    },
    {
      name: "og_image_url", // Match database column name
      type: "text",
      label: "OG Image URL",
      admin: {
        description: "OpenGraph image URL from meta tags",
      },
    },
    {
      name: "og_url", // Match database column name
      type: "text",
      label: "OG URL",
      admin: {
        description: "OpenGraph canonical URL",
      },
    },
    {
      name: "source_api_url", // Match database column name
      type: "text",
      label: "Source API URL",
      admin: {
        description: "Savee API endpoint for source resolution",
      },
    },
    {
      name: "saved_at", // Match database column name
      type: "text",
      label: "Saved At",
      admin: {
        description: "ISO timestamp when item was scraped",
      },
    },

    // Rich Metadata for Filtering/Search
    {
      name: "color_hexes", // Match database column name
      type: "json",
      label: "Color Hex Codes",
      admin: {
        description: "Array of hex color codes extracted from image",
      },
    },
    {
      name: "ai_tags", // Match database column name
      type: "json",
      label: "AI Generated Tags",
      admin: {
        description: "AI-generated descriptive tags for content",
      },
    },
    {
      name: "colors",
      type: "json",
      label: "RGB Colors",
      admin: {
        description: "Array of RGB color values",
      },
    },
    {
      name: "links",
      type: "json",
      label: "Sidebar Links",
      admin: {
        description: "Links extracted from item sidebar",
      },
    },
    {
      name: "metadata",
      type: "json",
      label: "Additional Metadata",
      admin: {
        description: "Complete sidebar info and other metadata",
      },
    },

    // Storage
    {
      name: "r2_key", // Match database column name
      type: "text",
      label: "R2 Storage Key",
      admin: {
        readOnly: true,
      },
    },

    // Error Handling
    {
      name: "errorMessage",
      type: "textarea",
      label: "Error Message",
      admin: {
        condition: (data) => data.status === "error",
      },
    },
  ],
};
